[FSDP2] Added `_get_managed_modules` for init

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118298
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* #117877
* #117867
* #117861
* __->__ #117852
* #117817
* #117814
* #117776
* #117984
* #118186
* #117994

This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.

---

This PR adds `_get_managed_states(managed_modules)` to return the managed parameters and buffers given the managed modules.
- Without an extra mechanism to ignore specific parameters or buffers, the rule currently is simply to get the directly managed state (i.e. parameters/buffers) from each managed module while de-duplicating shared ones.
- However, we prefer this translation from managed modules to managed states to accommodate ignoring specific states in the future (which has appeared in various open-source use cases).


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
[FSDP2] Added `_get_managed_modules` for init

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added `_get_managed_modules` for init"


This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added `_get_managed_modules` for init"


This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added `_get_managed_modules` for init"


This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_common.py:
```diff
@@ -49,3 +51,11 @@ class HSDPMeshInfo(FSDPMeshInfo, DDPMeshInfo):
+    if registry is None:
+        return True
+    # TODO: Add the TorchRec composable API name.
+    return "replicate" not in registry
```
this is reading registry. I guess somewhere we are writing to registry with ``registry["replicate"] = `` ? should we enforce the string/key consistentcy with CONSTANT_REPLICATE or some enum?

awgu replied:
The string is by the function name, so it is always `"replicate"` for the function `replicate`.
https://github.com/pytorch/pytorch/blob/c6be5d55a56cc12b7a004acdb6a7da92ee2142f7/torch/distributed/_composable/contract.py#L97

I guess we could use `replicate.__name__` here as well, but I think it will always be `"replicate"`.

weifengpy replied:
got you. I like ``replicate.__name__`` - it saved my time for guessing whether string ``replicate`` is arbitrary

awgu replied:
I guess one other consideration is that then, every composable API needs to import every other composable API to get the function (e.g. `replicate`) to be able to ignore it.

This system overall is not super robust, so I wonder what a better approach would be. For example, we will want to ignore the TorchRec API, but we may not be able to import their function. Using the string directly might be simpler. I can add a comment.

fegin replied:
We can not do `replicate.__name__` because `replicate` is going to do the same check as well. This will lead to a circular dependency. Just some random though, maybe we can define an enum and also have a util API in composable to help with the check?

Commit:
Update on "[FSDP2] Added `_get_managed_modules` for init"


This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented:
> except those already managed by

just being explicit, do we ban mixing old and new APIs (what if fully_shard() encounters an FSDP wrapper or a DDP wrapper)?

awgu commented:
> > except those already managed by
> 
> just being explicit, do we ban mixing old and new APIs (what if fully_shard() encounters an FSDP wrapper or a DDP wrapper)?

Ah, good point. We do not ban it, but the results will be wrong. `fully_shard` here will not know to stop recursing into the FSDP or DDP wrappers.

wanchaol commented on torch/distributed/_composable/fsdp/_fsdp_init.py:
```diff
@@ -28,3 +34,27 @@ def _init_default_fully_shard_mesh(device_type: str) -> DeviceMesh:
+    # Track visisted modules to avoid visiting shared modules multiple times
+    visited_modules: Set[nn.Module] = set()
+
+    def dfs(module: nn.Module) -> None:
```
I feel we can just loop through the `root_module.named_modules()` API where it already gives a dfs order? we don't need to keep the visited module logic if we just reuse that

awgu replied:
If we use `named_modules()`, since that is a generator, I am not sure how to skip the child modules of a `replicate` module. (E.g., there could be modules after the `replicate` subtree after that we should not skip.)

wanchaol commented on torch/distributed/_composable/fsdp/_fsdp_common.py:
```diff
@@ -49,3 +51,11 @@ class HSDPMeshInfo(FSDPMeshInfo, DDPMeshInfo):
+    if registry is None:
+        return True
+    # TODO: Add the TorchRec composable API name.
+    return "replicate" not in registry
```
To confirm, this means FSDP2 would not step into `replicate` submodule, but it would step into all other submodules?

awgu replied:
Yep!

wanchaol reviewed (approved):
Only some nits about how to loop through managed modules

Commit:
Update on "[FSDP2] Added `_get_managed_modules` for init"


This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added `_get_managed_modules` for init"


This PR adds `_get_managed_modules()` to determine which modules a `fully_shard(module)` call manages. The rule is defined as:
> `fully_shard(module)` manages all modules in `module.modules()` except those already managed by a nested `fully_shard()` or a nested non-composable API (e.g. `replicate()` or TorchRec).

Practically, this can be implemented as a graph search from `module` that does not proceed into any module with `fully_shard` or a non-composable API applied. Because the non-composable APIs follow the same rule, this rule is correct inductively.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fegin commented on torch/distributed/_composable/fsdp/_fsdp_init.py:
```diff
@@ -28,3 +34,50 @@ def _init_default_fully_shard_mesh(device_type: str) -> DeviceMesh:
+    visited_params: Set[nn.Parameter] = set()
+    visited_buffers: Set[torch.Tensor] = set()
+    for module in modules:
+        for param_name, param in module.named_parameters(recurse=False):
```
The name does not seem to be used.

fegin commented on torch/distributed/_composable/fsdp/_fsdp_init.py:
```diff
@@ -28,3 +34,50 @@ def _init_default_fully_shard_mesh(device_type: str) -> DeviceMesh:
+            return  # nested `fully_shard` module
+        visited_modules.add(module)
+        for submodule in module.children():
+            if submodule is not None and submodule not in visited_modules:
```
Just out of curiosity, when will `children()` return None?

awgu replied:
Good point. In normal PyTorch, I think it cannot return `None` since `_named_members()` filters `None`. I am not sure about internal overrides. We can remove this check until we hit the error.

fegin commented on torch/distributed/_composable/fsdp/_fsdp_init.py:
```diff
@@ -28,3 +34,50 @@ def _init_default_fully_shard_mesh(device_type: str) -> DeviceMesh:
+                continue
+            params.append(param)
+            visited_params.add(param)
+        for buffer_name, buffer in module.named_buffers(recurse=False):
```
ditto

awgu commented:
Squashed into preceding PR
